src/ring_theory/polynomial.lean

```diff
@@ -47,6 +49,84 @@ begin
+λ i, finsupp.mem_support_iff.trans (not_iff_not_of_iff ⟨λ H, subtype.eq H, subtype.mk.inj⟩)⟩
+
+@[simp] theorem coeff_restriction {p : polynomial R} {n : ℕ} : ↑(coeff (restriction p) n) = coeff p n := rfl
+
+@[simp] theorem coeff_restriction' {p : polynomial R} {n : ℕ} : (coeff (restriction p) n).1 = coeff p n := rfl
+
+@[simp] theorem degree_restriction {p : polynomial R} : (restriction p).degree = p.degree := rfl
+
+@[simp] theorem nat_degree_restriction {p : polynomial R} : (restriction p).nat_degree = p.nat_degree := rfl
+
+@[simp] theorem monic_restriction {p : polynomial R} : monic (restriction p) ↔ monic p :=
+⟨λ H, congr_arg subtype.val H, λ H, subtype.eq H⟩
+
+@[simp] theorem restriction_zero : restriction (0 : polynomial R) = 0 := rfl
+
+@[simp] theorem restriction_one : restriction (1 : polynomial R) = 1 :=
+ext.2 $ λ i, subtype.eq $ by rw [coeff_restriction', coeff_one, coeff_one]; split_ifs; refl
+
+variables {S : Type v} [comm_ring S] {f : R → S} {x : S}
+
+theorem eval₂_restriction {p : polynomial R} :
+  eval₂ f x p = eval₂ (f ∘ subtype.val) x p.restriction :=
+rfl
+
+section to_subring
+variables (p : polynomial R) (T : set R) [is_subring T]
+
+/-- Given a polynomial `p` and a subring `T` that contains the coefficients of `p`,
+return the corresponding polynomial whose coefficients are in `T. -/
+def to_subring (hp : ↑p.frange ⊆ T) : polynomial T :=
+⟨p.support, λ i, ⟨p.to_fun i,
+  if H : p.to_fun i = 0 then H.symm ▸ is_add_submonoid.zero_mem _
+  else hp $ finsupp.mem_frange.2 ⟨H, i, rfl⟩⟩,
+λ i, finsupp.mem_support_iff.trans (not_iff_not_of_iff ⟨λ H, subtype.eq H, subtype.mk.inj⟩)⟩
+
+variables (hp : ↑p.frange ⊆ T)
+include hp
+
+@[simp] theorem coeff_to_subring {n : ℕ} : ↑(coeff (to_subring p T hp) n) = coeff p n := rfl
+
+@[simp] theorem coeff_to_subring' {n : ℕ} : (coeff (to_subring p T hp) n).1 = coeff p n := rfl
+
+@[simp] theorem degree_to_subring : (to_subring p T hp).degree = p.degree := rfl
+
+@[simp] theorem nat_degree_to_subring : (to_subring p T hp).nat_degree = p.nat_degree := rfl
+
+@[simp] theorem monic_to_subring : monic (to_subring p T hp) ↔ monic p :=
+⟨λ H, congr_arg subtype.val H, λ H, subtype.eq H⟩
+
+omit hp
+
+@[simp] theorem to_subring_zero : to_subring (0 : polynomial R) T (set.empty_subset _) = 0 := rfl
+
+@[simp] theorem to_subring_one : to_subring (1 : polynomial R) T
+  (set.subset.trans (finset.coe_subset.2 finsupp.frange_single)
+    (set.singleton_subset_iff.2 (is_submonoid.one_mem _))) = 1 :=
+ext.2 $ λ i, subtype.eq $ by rw [coeff_to_subring', coeff_one, coeff_one]; split_ifs; refl
+end to_subring
+
+variables (T : set R) [is_subring T]
+
+/-- Given a polynomial whose coefficients are in some subring, return
+the corresponding polynomial whose coefificents are in the ambient ring. -/
+def of_subring (p : polynomial T) : polynomial R :=
+⟨p.support, subtype.val ∘ p.to_fun,
+λ n, finsupp.mem_support_iff.trans (not_iff_not_of_iff
+  ⟨λ h, congr_arg subtype.val h, λ h, subtype.eq h⟩)⟩
+
```

Comment:
This is just `map subtype.val`, I don't think it needs to be defined again.

Reply:
This has better definitional equalities.